pkg/storage/cacher.go

```diff
@@ -859,10 +859,19 @@ func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
 	case !curObjPasses && oldObjPasses:
 		watchEvent = watch.Event{Type: watch.Deleted, Object: object}
 	}
+
+	// We don't want to block on c.result if c.done is already closed.
+	//
+	// TODO: This doesn't protect with the situation when c.result is
+	// already full and we are handing on it and in the meantime c.done
+	// is being closed and noone will clear the result channel.
+	// However, we can't heve select of two cases, as this may lead to
+	// delivering random part of events after c.done is closed but we
+	// are still sending them for a moment.
```

Comment:
>I understand it's non-deterministic, but I still fail to see why it's a problem. If the event is not sent to the client, client will see the event in its next Watch request, right?

@caesarxuchao assuming that any event after this one also won't be delivered then yes that's the case. However, the client is starting the next watch from the last even it has seen. That said if we deliver to him events 1 and 3 (and drop 2), it will start the next watch from 4.

That said, the real problem is that when "c.done" is close we have completely no control and guarantee whever the items in the ResultChan will or will not be delivered to the watcher.